specs/chugsplash-manager.md

```diff
@@ -0,0 +1,178 @@
+# ChugSplashManager Speed Spec
+
+## Bundle states
+
+```typescript
+enum ActionType {
+  SET_CODE,
+  SET_STORAGE
+}
+```
+
+```typescript
+enum BundleStatus {
+  EMPTY,
+  PROPOSED,
+  APPROVED,
+  CANCELLED,
+  COMPLETED
+}
+```
+
+```typescript
+struct BundleState {
+  BundleStatus status;
+  bool[] executions;
+  uint256 actionsExecuted;
+  uint256 timeClaimed;
+  address selectedExecutor;
+}
+```
+
+```typescript
+struct Action {
+  string target;
+  ActionType actionType;
+  uint256 gasUsed;
+  bytes code;
+}
+```
+
+## State variables
+
+`EXECUTION_COST_MULTIPLIER`: Ensure that enough ETH is deposited in the ChugSplashManager at the time that a bundle is approved. This allows a bundle to be executed even if the gas price spikes while it is active.
+
+## Computing a unique bundle ID
+
+```typescript
+const computeBundleId = (
+  bundleRoot: bytes32,
+  bundleSize: uint256,
+  configUri: string
+): bytes => {
+  return keccak256(
+    abi.encode(
+      bundleRoot,
+      bundleSize,
+      configUri,
+      totalGasUsed
+    )
+  )
+}
+```
+
+## Getting the selected executor
+
+```typescript
+const getSelectedExecutor = (
+  bundleId: bytes32
+): address => {
+  return bundles[bundleId].selectedExecutor
+}
+```
+
+## Getting the proxy by name
+
+```typescript
+const getProxyByName = (
+  name: string
+): address payable => {
+  return (
+    payable(
+      Create2.compute(address(this), keccak256(bytes(name)), type(Proxy).creationCode)
+    )
+  )
+}
+```
+
+## Proposing a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be proposable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST be in the `EMPTY` state before they can be proposed.
+* Proposing a bundle puts the bundle in the `PROPOSED` state.
+* The ChugSplashManager MUST emit the event ChugSplashBundleProposed with the ID of the bundle, the bundle root, the bundle size, and the config URI as parameters. The ChugSplashManager MUST announce this event to the registry.
+
+## Approving a ChugSplash bundle
+
+* A ChugSplash bundle MUST only be approvable by the `admin` of the `ChugSplashManager`.
+* Bundles MUST have be in the `PROPOSED` state before they can be approved.
+* There MUST NOT be any active bundle.
+* Approving a bundle MUST put the bundle in the `APPROVED` state.
+* Approving a bundle MUST make the approved bundle the active bundle.
+* The amount of ETH available in the ChugSplashManager to cover the cost of deployment MUST be greater than or equal to the estimated upper bound for the cost of the deployment. In other words, `address(this).balance - totalDebt - BUNDLE_EXECUTION_BOND_AMOUNT >= EXECUTION_COST_MULTIPLIER * totalGasUsed * tx.gasPrice`.
```

Comment:
I felt like this was a reasonable approach since we don't know how much a deployment will cost up-front due to fluctuations in gas price. So, the purpose of this line is to ensure that the owner has deposited more than enough ETH even if the gas price spikes while the bundle is being executed.

Reply:
Hmmm this one is interesting. I think we should decide based on the ideal user experience. I do like the idea of estimating ahead of time what the upper-bound cost will be. I'm not entirely convinced that enforcing this at the smart contract level is best though. My preference is still to only require the bond up-front, but have the UI estimate how much you will need to put into the account for someone to pick up the job.

Reply:
Sounds good, I changed this back to what it was before